Add user identity to cluster status to avoid leakage when switching account

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

Describe the changes in this PR:

With #1489, when a user switches account, `sky status -r` will remove the previously launched cluster from the table due to the cluster cannot be found in the new account, causing resource leakage. It could happen whenever user switches account on any cloud. This is to get rid of the problem.

---
IMPORTANT: A user with multiple identities for a cloud need to do the following things after upgrading to this PR:
1. Switch to the account used for launching the clusters in the cluster table
2. `sky status -r` so that the owner information can be correctly updated in our cluster table cache.

A normal user will not be affected by the PR.

---

<!-- Describe tests ran in a "Tested:" section -->
<!-- Unit tests (tests/test_*.py) are part of Github CI; the below are additional tests. -->

Tested (run the relevant ones):
- [x] `sky launch -c min --cloud aws`; switch to another account; `sky status -r`
<img width="1284" alt="image" src="https://user-images.githubusercontent.com/6753189/206991216-8661b713-378d-4044-a2d8-739166be5046.png">

- [x] `sky launch -c min --cloud aws`; switch to another account; `sky launch -c min`; `sky down min`; `sky stop min`; `sky start min`
- [x] `sky launch -c min --cloud gcp`; switch to another account; `sky status -r`
<img width="1352" alt="image" src="https://user-images.githubusercontent.com/6753189/206994122-c112a12c-12ba-4138-9ece-4eb8f7804349.png">
<img width="1306" alt="image" src="https://user-images.githubusercontent.com/6753189/206994478-467ef401-fad0-4e73-952e-63b844b2bb37.png">

- [x] `sky launch -c min --cloud azure`; switch to another account; `sky status -r`
<img width="1863" alt="image" src="https://user-images.githubusercontent.com/6753189/206998731-7afab82f-6cc0-4a06-94f4-8483c33bf9f8.png">

- [x] `./tests/run_smoke_tests.sh` (except for TPU VM tests #1500 and list source files #1510).
- [x] `./tests/backward_compatibility.sh`

Michaelvll commented:
Testing the smoke tests again with 4df9d9e

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -189,6 +189,15 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
         raise NotImplementedError
 
+    def get_cloud_user_identity(self) -> Optional[str]:
+        """Returns the identity of the user on this cloud.
```
Can we enrich this comment? Since "identity" is a term we coin. (Clouds do have this term but have different meanings).

Some ideas:
```
(Advanced) Returns the currently active identity of the user on this cloud.

The user "identity" is associated with each SkyPilot cluster he/she creates. This is used in protecting cluster operations in a multi-identity scenario, where the same user/device can switch between different cloud identities.

The choice of what constitutes an identity is up to each cloud's implementation. In general, to suffice for the above purposes, ensure that different identities imply different resources are displayed in the cloud's console.

Example identities (see cloud implementations):
  - AWS: ....
  - GCP: email address + project ID
  - Azure: email address + subscription ID
  - <new cloud>: ...


Returns:
  None iff ...; otherwise the current user identity.
```

concretevitamin commented on sky/clouds/gcp.py:
```diff
@@ -457,6 +459,22 @@ def get_credential_file_mounts(self) -> Dict[str, str]:
+        if not account:
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.CloudUserIdentityError(
+                    'No GCP account is activated.')
```
add: Try running `gcloud auth list --filter=status:ACTIVE --format="value(account)"` and ensure it correctly returns the current user.

concretevitamin commented on sky/clouds/gcp.py:
```diff
@@ -457,6 +459,22 @@ def get_credential_file_mounts(self) -> Dict[str, str]:
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.CloudUserIdentityError(
+                    'No GCP account is activated.')
+        return f'{account}@{self.get_project_id()}'
```
It's confusing to have two `@`, especially if these strings can get displayed. How about `a@b.com [project_id={project_id}]`?

concretevitamin commented on sky/clouds/azure.py:
```diff
@@ -278,13 +279,15 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
-        # If Azure is properly logged in, this will return something like:
-        #   {"id": ..., "user": ...}
+        # If Azure is properly logged in, this will return the account email
+        # address.
```
L284-285 remnant?

concretevitamin reviewed (commented):
Thanks @Michaelvll. Read everything except backend_utils.py. Leaving some comments first.

Michaelvll commented on sky/global_user_state.py:
```diff
@@ -255,6 +263,25 @@ def set_cluster_metadata(cluster_name: str, metadata: Dict[str, Any]) -> None:
+                              user_identity: Optional[str]) -> None:
+    if user_identity is None:
+        return
+    # This must be called when the lock is acquired for the cluster.
+    metadata = get_cluster_metadata(cluster_name)
+    if metadata is None:
+        metadata = {}
+    metadata['user_identity'] = user_identity
+    set_cluster_metadata(cluster_name, metadata)
```
This is not applicable, as we are currently using a lock for the whole cluster, which will lock the whole process of `sky launch`, status refreshes, etc that operates on that cluster. In that case, having another lock for this does not seem necessary to me and may introduce additional overhead?

Michaelvll replied:
After discussing offline with @concretevitamin, we found that our current implementation will acquire the lock for the cluster, even if we don't need to refresh the status. That is not as expected, especially for the operations that does not require lock before, such as `queue`, `exec`.

Michaelvll commented on sky/templates/spot-controller.yaml.j2:
```diff
@@ -39,6 +39,8 @@ run: |
 envs:
   SKYPILOT_USAGE_USER_ID: {{logging_user_hash}}
+  # skip cloud identity check for spot controller to avoid the overhead.
+  SKYPILOT_SKIP_CLOUD_IDENTITY_CHECK: 1
```
This is retrieved at the beginning of `backend_utils._check_user_identity_no_lock()`, so that the spot controller will skip the checking of the user identity

Michaelvll commented on sky/clouds/gcp.py:
```diff
@@ -470,27 +488,10 @@ def accelerator_in_region_or_zone(self,
-        # `gcloud info --format='value(config.project)'`
         if dryrun:
             return 'dryrun-project-id'
-        if 'GOOGLE_APPLICATION_CREDENTIALS' in os.environ:
-            gcp_credential_path = os.environ['GOOGLE_APPLICATION_CREDENTIALS']
-        else:
-            gcp_credential_path = DEFAULT_GCP_APPLICATION_CREDENTIAL_PATH
-        if not os.path.exists(gcp_credential_path):
-            with ux_utils.print_exception_no_traceback():
-                raise FileNotFoundError(
-                    f'No GCP credentials found at '
-                    f'{gcp_credential_path}. Please set the '
-                    f'GOOGLE_APPLICATION_CREDENTIALS '
-                    f'environment variable to point to '
-                    f'the path of your credentials file.')
-
-        with open(gcp_credential_path, 'r') as fp:
-            gcp_credentials = json.load(fp)
-        project_id = gcp_credentials.get('quota_project_id',
-                                         None) or gcp_credentials['project_id']
```
The `auth.default()` is official method that returns the credential and project id, which retrieves the project id just as we did before. I think it should be cleaner to use the official method instead of doing it by our own?
https://google-auth.readthedocs.io/en/master/reference/google.auth.html
https://google-auth.readthedocs.io/en/master/_modules/google/auth/_default.html#default

Michaelvll commented on sky/clouds/azure.py:
```diff
@@ -305,6 +308,23 @@ def accelerator_in_region_or_zone(self,
+        except (FileNotFoundError, knack.util.CLIError, ImportError):
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.CloudUserIdentityError(
+                    'Failed to get activated Azure account.') from None
```
Good point! Removed the ImportError.

Michaelvll replied:
Yes, it is part of the dependencies for `skypilot[azure]`. Tried with `docker run -it --rm continuumio/miniconda3`, and `python -c 'import knack'`  works after installing `skypilot[azure]`

Michaelvll commented on sky/clouds/aws.py:
```diff
@@ -319,22 +322,47 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
-            return False, (
-                'Failed to access AWS services with credentials stored in ~/.aws/credentials.'
-                ' Make sure that the access and secret keys are correct.'
-                ' To reconfigure the credentials, ' + help_str[1].lower() +
-                help_str[2:])
```
Our previous code was to make the first charactor lower case, as we were using comma. Now it is after a period, so no need to make the first charactor lower case.

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1600,8 +1601,32 @@ def _query_status_azure(
+    user_identity = global_user_state.get_cluster_user_identity(cluster_name)
+    current_user_identity = cloud.get_cloud_user_identity()
+    if user_identity is None:
+        if current_user_identity is not None:
```
This can be `None`, if we fail to retrieve the user identity. In that case, we don't update the user identity for the cluster.

Michaelvll commented:
We now check the identity only when the code modifies the cluster or has some cloud API operations. Including the following places:
1. Before provisioning on an existing cluster;
2. Before stopping/terminating a cluster;
3. Before refreshing the status of a cluster.

Other operations that go through `ssh` will not be checked so that users can still interact with those clusters after switching the cloud account. (If the cluster is a spot instance or has autostop/setup, most operations will still fail, as the status refresh will be triggered before them.)

We also made the cluster owner a column of the cluster table to utilize the database transaction instead of our per-cluster lock, which was too coarse-grained.

PTAL @concretevitamin


TODO:
- [ ] Make sure after log in again for a expired SSO account will have the same UserID.
- [ ] `tests/run_smoke_tests.sh`
- [ ] `tests/backward_compatibility_tests.sh`

concretevitamin commented on sky/clouds/aws.py:
```diff
@@ -289,78 +305,79 @@ def _make(instance_list):
-                # TODO(zhwu): after we publish sky to PyPI,
-                # change this to `pip install sky[aws]`
+                ' Run the following commands in the SkyPilot codebase:'
                 '\n     $ pip install .[aws]'
```
Can this be changed to `pip install "skypilot[aws]"` now?

Michaelvll replied:
Should we do it when we publish the new release including these changes? Added a TODO here to change the installation command. Wdyt?

concretevitamin commented on sky/clouds/aws.py:
```diff
@@ -289,78 +305,79 @@ def _make(instance_list):
+            # 'Arn' is the full path to the user, which can be reused when
             # the user is deleted and recreated.
+            # Refer to https://docs.aws.amazon.com/cli/latest/reference/sts/get-caller-identity.html # pylint: disable=line-too-long
+            # and https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html # pylint: disable=line-too-long
```
Looks like 2nd link can be replaced with https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html#principaltable?

Michaelvll replied:
Good point! Replaced. Thanks

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1607,8 +1608,56 @@ def _query_status_azure(
+            f'Failed to get the current user identity for {cloud}: {e}')
+
+
+def check_owner_identity(cluster_name: str):
```
nit: -> None

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1607,8 +1608,56 @@ def _query_status_azure(
+    owner_identity = record['owner']
+    if current_user_identity is None:
+        # Skip the check if the cloud does not support user identity,
+        # or we fail to get the user identity.
```
When we fail to get identity, why is it ok to let this operation go through? See another comment about potential leakage.

In the case of "if the cloud does not support user identity" it may be ok; likely should check any recorded owner = None.

Michaelvll replied:
I was intended to make this check minimum, so that it just serves as a minimum safeguard, but not panic for failing to get the owner identity.
However, it makes sense that this can cause leakage. Let me revert it to using the `get_current_user_identity` that will raise error (also for the one in `backend.provision()`).

concretevitamin replied:
Sounds good; should we remove L1663?

Michaelvll replied:
Removed. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1715,7 +1764,9 @@ def _update_cluster_status_no_lock(
         cluster_name: str,
         acquire_per_cluster_status_lock: bool) -> Optional[Dict[str, Any]]:
-    """Update the cluster status by checking ray cluster and real status from cloud.
+    """Update the cluster identity and status.
```
```suggestion
    """Update the cluster status and owner identity.
```

Michaelvll replied:
I changed the docstr a bit to `Check/update owner identity and update the cluster status.`. Wdyt?

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1756,7 +1809,8 @@ def refresh_cluster_status_handle(
-        if force_refresh or record['autostop'] >= 0:
+        use_spot = handle.launched_resources.use_spot
+        if (force_refresh or record['autostop'] >= 0 or use_spot):
             # Refresh the status only when force_refresh is True or the cluster
             # has autostopped turned on.
```
Remove? Since it's textual description of the conditions.

Michaelvll replied:
Good point. Removed. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1725,6 +1776,8 @@ def _update_cluster_status(
       If the cluster is terminated or does not exist, return None.
       Otherwise returns the input record with status and ip potentially updated.
     """
```
Add a Raises:?

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1845,13 +1903,22 @@ def _refresh_cluster(cluster_name):
 
     # Show information for removed clusters.
-    autodown_clusters, remaining_clusters = [], []
+    normal_updated_records = []
```
kept_records?

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/core.py:
```diff
@@ -205,6 +205,8 @@ def start(
         sky.exceptions.NotSupportedError: if the cluster to restart was
           launched using a non-default backend that does not support this
           operation.
+        sky.exceptions.ClusterOwnerIdentitiesMismatchError: if the cluster to
+            restart was launched by a different user.
```
Is this from `backend_utils.refresh_cluster_status_handle()`? There are other places in this module that call this.

Michaelvll replied:
It is from the `refresh_cluster_status_handle` (which will only be called when the cluster has autostop set up or is a spot cluster) or the `backend.provision` (we check the identity before provisioning the cluster.)

concretevitamin replied:
I meant that: should we update the `Raises:` section for all code paths that may raise? 
- `autostop()`
- `queue/cancel/tail_logs/download_logs/job_status` -> `_check_cluster_available` -> `backend_utils.refresh_cluster_status_handle`
- `spot_*()` -> `_is_spot_controller_up` -> `backend_utils.refresh_cluster_status_handle`

Michaelvll replied:
That is a great point! In the latest commit, I catch the error when the `refresh_cluster_status_handle` is called for those operations so that they will not be blocked by the identity check.

concretevitamin commented on sky/global_user_state.py:
```diff
@@ -30,6 +30,8 @@
 
 
 def create_table(cursor, conn):
+    # Enable WAL mode to avoid locking issues.
```
Link to issue/PR number?

Michaelvll replied:
Good point! Added. Thanks!

concretevitamin commented on sky/global_user_state.py:
```diff
@@ -263,31 +275,32 @@ def set_cluster_metadata(cluster_name: str, metadata: Dict[str, Any]) -> None:
-    if metadata is None:
-        return None
-    return metadata.get('user_identity')
+def get_owner_identity_for_cluster(cluster_name: str) -> Optional[str]:
```
`ag` shows unused.

Michaelvll replied:
Ahh, good catch! We directly get it from the `record` we get using `get_cluster`. Removed it. Thanks!

concretevitamin reviewed (commented):
Thanks @Michaelvll. Did a pass; haven't tried it out yet.

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -189,9 +189,27 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
+        implementation. In general, to suffice for the above purposes,
+        ensure that different identities imply different resources are
+        displayed in the cloud's console.
+
```
> Is it ok for a new identity to autostop a cluster?

I have thought about it. I think it should be fine to allow a new identity to autostop the cluster because:
1. The cluster is all launched by the user. I think the user should have the maximum permission to operate the cluster.
2. Autostop will work, as the cluster has the correct identity and can stop itself.

concretevitamin replied:
Make sense to me.

Can we add (1) when is identity check performed
> We now check the identity only when the code modifies the cluster or has some cloud API operations. Including the following places:
> 
> 1. Before provisioning on an existing cluster;
> 2. Before stopping/terminating a cluster;
> 3. Before refreshing the status of a cluster.
> 
> Other operations that go through `ssh` will not be checked so that users can still interact with those clusters after switching the cloud account. (If the cluster is a spot instance or has autostop/setup, most operations will still fail, as the status refresh will be triggered before them.)

and (2) this comment about the design choice of "the user should have the maximum permission to operate the cluster." + "as long as it'll work, we allow it" here?

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1630,19 +1637,22 @@ def _check_user_identity_no_lock(cluster_name: str):
-        global_user_state.set_cluster_user_identity(cluster_name,
-                                                    current_user_identity)
-    elif user_identity != current_user_identity:
+    # The user identity can be None, if the cluster is created by an older
+    # version of SkyPilot. In that case, we set the user identity to the
+    # current one.
+    if owner_identity is None:
+        global_user_state.set_owner_identity_for_cluster(
+            cluster_name, current_user_identity)
```
We should ask the user to run `sky status -r` immediately after updating to this PR. Otherwise, it is hard to get the correct information.

concretevitamin replied:
Sounds good. Could we add this possible failure case here? Just for our own note.

Michaelvll replied:
Added. Thanks!

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1600,8 +1602,49 @@ def _query_status_azure(
 }
 
 
+def get_cloud_user_identity_no_error(cloud: clouds.Cloud) -> Optional[str]:
```
Removed the method to avoid potential leakage. No need for the change here. : )

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1611,14 +1610,22 @@ def _query_status_azure(
+def check_owner_identity(cluster_name: str):
+    """Check if the current user is the same as the user who created the cluster.
+
+    This should be called before any operation that will modify the cluster, or query
```
Done. Thanks!

Michaelvll commented on sky/check.py:
```diff
@@ -41,7 +41,11 @@ def check(quiet: bool = False) -> None:
-             'and run ' + click.style('sky check', bold=True) + '.')
+             'and run ' + click.style('sky check', bold=True) + '.'
+             '\n' + click.style(
+                 'If any problem happen, please file an issue at '
```
Done. Thanks!

Michaelvll commented on sky/cli.py:
```diff
@@ -2271,6 +2274,8 @@ def _down_or_stop(name: str):
                     f'{colorama.Style.RESET_ALL}')
+            except exceptions.ClusterOwnerIdentityMismatchError as e:
+                message = str(e)
             except exceptions.NotSupportedError as e:
```
Done. Thanks!

Michaelvll commented on sky/cli.py:
```diff
@@ -1990,9 +1990,12 @@ def start(
                        force=force)
+        except exceptions.ClusterOwnerIdentityMismatchError as e:
+            click.echo(str(e))
         except exceptions.NotSupportedError as e:
```
Done. Thanks!

Michaelvll commented on sky/clouds/azure.py:
```diff
@@ -323,7 +321,7 @@ def get_cloud_user_identity(self) -> Optional[str]:
                     'Failed to get Azure user identity with unknown '
                     f'exception: {type(e)} {e}') from e
-        return f'{account_email}@{self.get_project_id()}'
+        return f'{account_email}[subscription={self.get_project_id()}]'
```
Good point! Done. Thanks

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -189,9 +189,27 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
-    def get_cloud_user_identity(self) -> Optional[str]:
-        """Returns the identity of the user on this cloud.
+    def get_current_user_identity(self) -> Optional[str]:
+        """(Advanced) Returns currently active user identity of on this cloud.
```
Done. Thanks!

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -189,9 +189,27 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
+        """(Advanced) Returns currently active user identity of on this cloud.
 
+        The user "identity" is associated with each SkyPilot cluster they
+        creates. This is used in protecting cluster operations in a multi-
```
It is also used to protect the `provision/teardown` as well. Added it to the comment. Thanks!

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -189,9 +189,27 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
+        The choice of what constitutes an identity is up to each cloud's
+        implementation. In general, to suffice for the above purposes,
+        ensure that different identities imply different resources are
+        displayed in the cloud's console.
```
Good point! Thanks!

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -189,9 +189,27 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
+            - Azure: email address + subscription ID
+
+        Returns:
+            None if the cloud does not have a concept of user identity;
```
Done. Thanks!

Michaelvll commented on sky/exceptions.py:
```diff
@@ -137,3 +137,8 @@ class InvalidClusterNameError(Exception):
+
+
+class ClusterOwnerIdentityMismatchError(Exception):
+    """The cluster owner identity does not match user identity."""
```
Done. Thanks!

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1705,10 +1748,12 @@ def _update_cluster_status_no_lock(
 
-def _update_cluster_status(
-        cluster_name: str,
+def _update_cluster_identity_and_status(
```
Sorry, that was a remnant. Revert it.

Michaelvll commented on sky/clouds/gcp.py:
```diff
@@ -472,8 +472,11 @@ def get_cloud_user_identity(self) -> Optional[str]:
+                    'auth list --filter=status:ACTIVE '
+                    '--format="value(account)"` and ensure it correctly '
+                    'returns the current user.')
+        return f'{account}[project={self.get_project_id()}]'
```
Good point! Changed to `project_id`. Thanks!

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1718,16 +1763,24 @@ def _update_cluster_status(
       Otherwise returns the input record with status and ip potentially updated.
     """
+
+    def apply_check():
```
This has been removed. No action needed. : )

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1718,16 +1763,24 @@ def _update_cluster_status(
+
+    def apply_check():
+        record = global_user_state.get_cluster_from_name(cluster_name)
+        _check_user_identity_no_lock(cluster_name)
```
This has been removed. No action needed. Thanks : )

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -190,26 +190,28 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
 
         Returns:
-            None if the cloud does not have a concept of user identity;
-            otherwise the currently active user identity.
```
keep the otherwise case?

Michaelvll replied:
Oops, it was accidentally removed. Added back. Thanks!

concretevitamin reviewed (commented):
I have these clusters in a non-SSO account:
```
» sst
Clusters
NAME   LAUNCHED        RESOURCES            STATUS   AUTOSTOP  COMMAND                        
dbg    29 secs ago     1x AWS(m6i.2xlarge)  UP       -         sky start dbg                  
jump3  20 hrs ago      1x AWS(t3.xlarge)    STOPPED  1m        sky launch -c jump3 -i1 -t...  
```

Then, in a new terminal, after `aws2 sso login --profile xxx` I ran
```
» AWS_PROFILE=xxxxxxx sst -r
W 12-17 10:40:59 backend_utils.py:1958] Cluster terminated on the cloud: dbg
W 12-17 10:40:59 backend_utils.py:1963] Failed to refresh status for 1 cluster:
W 12-17 10:40:59 backend_utils.py:1968] Cluster  Error                                         
W 12-17 10:40:59 backend_utils.py:1968] jump3    The cluster 'jump3' (on AWS) is created by the account 'abcd', but the activated one is '1234'.
Clusters
...
```
 `dbg` from the old account is unexpectedly removed from the cluster table. Confirmed that the console of the non-SSO account still shows it as running.

concretevitamin commented:
Potential issue - 
```
AWS_PROFILE=xxxx sky launch -t t2.micro -c sso -i0 --down
```
This launches `sso` in SSO account. (`~/.aws/credentials` gets uploaded to this node.)  It failed to be autodown'ed, and while it's trying to it launches a new node in my non-SSO account. Commit dc2a425b5.

(maybe this is fixed in the SSO PR?)

---

Potential issue - after launching in SSO account an `sso` cluster, I did (without AWS_PROFILE set)
```
» sky down sso
Terminating 1 cluster: sso. Proceed? [Y/n]:
Traceback (most recent call last):
...
  File "/Users/zongheng/Dropbox/workspace/riselab/sky-computing/sky/utils/common_utils.py", line 164, in read_yaml
    with open(path, 'r') as f:
<generated ray yaml file not found error>
```

(maybe this is fixed in the SSO PR?)

Michaelvll commented:
> `dbg` from the old account is unexpectedly removed from the cluster table. Confirmed that the console of the non-SSO account still shows it as running.

As discussed offline, this problem could be caused by not running `sky status -r` with the same account for launching the existing cluster immediately after upgrading to this branch. 

Here is an option @concretevitamin proposed: Adding a prompt for clusters having an empty `owner` and aborting the program to let the user update the status with `sky status -r`. However, we decide to stick wit the current way to avoid letting non-multi-account users (most of the users) swallow the UX burden.

> Potential issue -
> 
> ```
> AWS_PROFILE=xxxx sky launch -t t2.micro -c sso -i0 --down
> ```
> 
> This launches `sso` in SSO account. (`~/.aws/credentials` gets uploaded to this node.) It failed to be autodown'ed, and while it's trying to it launches a new node in my non-SSO account. Commit [dc2a425](https://github.com/skypilot-org/skypilot/commit/dc2a425b57db01fad304245bfa7876cc28c6c9b7).
> 
> (maybe this is fixed in the SSO PR?)

Yes, this one is solved in the SSO PR. This PR does not handle credential problems.

> Potential issue - after launching in SSO account an `sso` cluster, I did (without AWS_PROFILE set)
> 
> ```
> » sky down sso
> Terminating 1 cluster: sso. Proceed? [Y/n]:
> Traceback (most recent call last):
> ...
>   File "/Users/zongheng/Dropbox/workspace/riselab/sky-computing/sky/utils/common_utils.py", line 164, in read_yaml
>     with open(path, 'r') as f:
> <generated ray yaml file not found error>
> ```
> 
> (maybe this is fixed in the SSO PR?)
> 
> Followup: without the profile env var, I `sky down` another non-SSO cluster, and the down succeeded.

I could not reproduce the issue, but I found another problem when trying the process -- the `INSERT or REPLACE` statement in the following place will overwrite the `owner` column if we do not explicitly reset the column with the old value. Just fixed it in the latest 0cb6a45.

Michaelvll commented on tests/backward_comaptibility_tests.sh:
```diff
@@ -35,7 +35,7 @@ if [ "$start_from" -le 1 ]; then
 rm -r  ~/.sky/wheels || true
 which sky
-sky launch -y -c ${CLUSTER_NAME} examples/minimal.yaml
+sky launch -y -c ${CLUSTER_NAME} examples/minimal.yaml || true
```
It was for fixing the issue that switching back to master does not work, when a new column is added to the cluster table, but it is not sufficient. I reverted it, and will apply another method for backward compatibility tests.

Michaelvll commented:
Tested (cb40ca6):
- [ ] `tests/backward_compatibility_tests.sh` (with the `sqlite3 ~/.sky/state.db 'ALTER TABLE clusters DROP owner;'` after each `conda activate sky-back-compat-master`)

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1631,8 +1636,50 @@ def _query_status_azure(
+    elif owner_identity != current_user_identity:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterOwnerIdentityMismatchError(
+                f'The cluster {cluster_name!r} (on {cloud}) is created by the account '
+                f'{owner_identity!r}, but the activated one is {current_user_identity!r}.'
```
```suggestion
                f'Cluster {cluster_name!r} ({cloud}) is owned by account '
                f'{owner_identity!r}, but the currently activated account is {current_user_identity!r}.'
```

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/global_user_state.py:
```diff
@@ -129,8 +129,12 @@ def add_or_update_cluster(cluster_name: str,
     status = ClusterStatus.UP if ready else ClusterStatus.INIT
     _DB.cursor.execute(
         'INSERT or REPLACE INTO clusters'
+        # All the fields need to exist here, even if they don't need
```
Great catch!

concretevitamin commented on sky/global_user_state.py:
```diff
@@ -294,6 +304,9 @@ def get_cluster_from_name(
     for row in rows:
+        # Explicitly specify the number of fields to unpack, so that
+        # we can add new fields to the database in the future without
+        # breaking the previous code.
```
We should consider namedtuple in the future.

Michaelvll replied:
Added the comment below. Thanks!

Michaelvll replied:
Added the comment below. Thanks!

concretevitamin reviewed (commented):
Thanks @Michaelvll; this is awesome.

Should we edit the PR description for upgrade instructions? Mainly asking multi-identity users to run `sky status -r`; users who use a single identity in each cloud are not affected.

----

Potential issue
```
» AWS_PROFILE=AdministratorAccess-1234 sky queue
Fetching and parsing job queue...
....
sky.exceptions.ClusterOwnerIdentityMismatchError: The cluster 'jump3' (on AWS) is created by the account 'yyyyyyy', but the activated one is 'xxxxxx'.
```

Same error for 
- `AWS_PROFILE=AdministratorAccess-1234 sky spot queue`.
- `AWS_PROFILE=AdministratorAccess-1234 sky exec jump3 echo hi`.

Both `jump3` and the spot controller were auto-stopped.

Haven't tried other commands.

----

Potential issue
- Ensure a cluster is STOPPED, both local cluster table & console
- Switched to this branch (note: below commands have not set the profile env var)
- `sky status -r`
- `sky start jump3`

Observed in console that a new node is launched, the old stopped node is leaked.

I terminated the new node, switched back to the `jump` branch, and re-did the above commands; observed the correct node was restarted.

Is this expected?

Michaelvll commented on sky/clouds/aws.py:
```diff
@@ -38,6 +40,22 @@ class AWS(clouds.Cloud):
+
+    _SSO_CREDENTIAL_HELP_STR = (
+        'Run the following commands:'
+        '\n      $ awsv2 configure sso'
+        '\n      $ awsv2 sso login --profile <profile_name>'
```
This is tricky since skypilot will install `awscli` as the dependency. The `aws` in the path can be in `v1`, which will fail for the `configure sso` and `sso login` (only supported by `v2`).

concretevitamin replied:
How about e.g.,
```
$ aws configure sso  # Must use aws v2 CLI
...
```
The reason is users may wrongly google/look for a tool called "awsv2"?

Michaelvll replied:
Good point! Added the `must use aws v2 cli` to the previous line.

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1659,12 +1659,14 @@ def check_owner_identity(cluster_name: str) -> None:
     # The user identity can be None, if the cluster is created by an older
     # version of SkyPilot. In that case, we set the user identity to the
     # current one.
+    # NOTE: a user upgrades SkyPilot and switches to a new cloud identity
+    # immediately without `sky status --refresh` first, can cause a leakage
+    # of the existing cluster.
```
Done. Thanks!

Michaelvll commented:
> Should we edit the PR description for upgrade instructions? Mainly asking multi-identity users to run sky status -r; users who use a single identity in each cloud are not affected.

Good call! Just added it to the description

> Error when doing operations on the cluster with auto-stop not requiring identity match.

Great catch!  We should allow those operators even if the autostop is set up. Fixed by catching the exceptions raised by cluster status refreshing for the operations that do not need the identity match.

The following now works:
```
sky launch -c min 'echo hi; sleep 10000000; echo bye'
sky autostop -i 100000 min  
AWS_PROFILE=admin-user sky logs min
AWS_PROFILE=admin-user sky queue
AWS_PROFILE=admin-user sky autostop min -i 10
AWS_PROFILE=admin-user sky cancel min -a
```

> Observed in console that a new node is launched, the old stopped node is leaked.

The problem may not be related to this PR as we don't change the generated ray yaml for the cluster. Do you think it is because the `jump` branch contains some additional settings in the ray yaml that is different from the one generated by the current branch causing the hash difference and resource leakage?

Michaelvll commented:
Tested:
- [ ] `tests/run_smoke_tests.sh`

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -507,3 +507,37 @@ def load_job_table_cache() -> Tuple[str, str]:
+            controller_status, handle = record['status'], record['handle']
+
+    if controller_status is None:
+        print('No managed spot job has been run.')
```
```suggestion
        print('No managed spot jobs found.')
```
Reason: it is possible a user ran a bunch of jobs, then `sky down` the controller. The new message seems to cover this case and the first-ever case together.

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -507,3 +507,37 @@ def load_job_table_cache() -> Tuple[str, str]:
+            msg += f'\n{stopped_message}'
+        if controller_status == global_user_state.ClusterStatus.INIT:
+            msg += '\nPlease wait for the controller to be ready.'
+        print(msg)
```
(For future: Sharing logging between the CLI (which doesn't show logging prefixes) and the programmatic API (which does) is a bit tough. Not sure what the best way is off the top of my head.)

Michaelvll replied:
Good point! I think one way is to use a logger handler that do not have the prefix for the CLI but the current one for the programmatic API. Will file an issue for this.

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1839,14 +1855,87 @@ def refresh_cluster_status_handle(
+                    cluster_name))
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterNotUpError(
+                f'{colorama.Fore.YELLOW}{operation} cluster {cluster_name!r} '
```
How about 
```
f'{colorama.Fore.YELLOW}{operation} cluster {cluster_name!r} is not allowed due to its status:'
f' {cluster_status.value}... skipped.'
...
```

concretevitamin replied:
See the other comment.

concretevitamin commented on sky/execution.py:
```diff
@@ -437,14 +437,8 @@ def exec(  # pylint: disable=redefined-builtin
-            raise ValueError(f'Cluster {cluster_name!r} is not up.  '
-                             'Use `sky status` to check the status.')
+    handle = backend_utils.check_cluster_available(cluster_name,
+                                                   'executing task on')
```
1. Other operation strings are missing "on"; some need an "of" such as: 
```
» sky queue
...
getting the job queue cluster 'jump3' (status: STOPPED)... skipped.
getting the job queue cluster 'smoke' (status: STOPPED)... skipped.
```
Seems a bit brittle to get the grammar right.

How about
```
Getting the job queue: skipped for cluster 'jump3' (status: STOPPED).
Executing tasks: skipped for cluster 'jump3' (status: STOPPED).
...
```
by passing the following?
```
operation='Getting the job queue'
operation='Executing tasks'
...
```

2. task -> tasks

Michaelvll replied:
That is a great point! Modified accordingly. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1839,14 +1855,87 @@ def refresh_cluster_status_handle(
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterNotUpError(
+                f'{colorama.Fore.YELLOW}Cluster {cluster_name!r} does not '
+                f'exist... skipped.{colorama.Style.RESET_ALL}')
```
```suggestion
                f'exist.{colorama.Style.RESET_ALL}')
```

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1644,7 +1644,9 @@ def check_owner_identity(cluster_name: str) -> None:
         exceptions.ClusterOwnerIdentityMismatchError: if the current user is not the
-            same as the user who created the cluster.
+          same as the user who created the cluster.
+        sky.exceptions.CloudUserIdentityError: if failed to get the activated user
```
```suggestion
        sky.exceptions.CloudUserIdentityError: if we fail to get the current user
```

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1800,9 +1803,11 @@ def _update_cluster_identity_and_status(
         sky.exceptions.ClusterOwnerIdentityMismatchError: the cluster's owner
-        identity mismatches the current cloud user.
+          identity mismatches the current cloud user.
+        sky.exceptions.CloudUserIdentityError: if failed to get the activated user
```
ditto

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1800,9 +1803,11 @@ def _update_cluster_identity_and_status(
     Raises:
         sky.exceptions.ClusterOwnerIdentityMismatchError: the cluster's owner
-        identity mismatches the current cloud user.
+          identity mismatches the current cloud user.
```
Keep the same as L1646-1647?

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1829,8 +1834,19 @@ def refresh_cluster_status_handle(
+    """Refresh the cluster status and return the status and handle.
+
+    Raises:
+        sky.exceptions.ClusterOwnerIdentityMismatchError: the cluster's owner
```
Keep the same as L1646-1647?

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1829,8 +1834,19 @@ def refresh_cluster_status_handle(
+    Raises:
+        sky.exceptions.ClusterOwnerIdentityMismatchError: the cluster's owner
+          identity mismatches the current cloud user.
+        sky.exceptions.CloudUserIdentityError: if failed to get the activated user
```
ditto

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1839,14 +1855,87 @@ def refresh_cluster_status_handle(
+    Raises:
+        exceptions.ClusterNotUpError: if the cluster is not UP.
+        exceptions.NotSupportedError: if the cluster is not based on
+          CloudVmRayBackend
```
```suggestion
          CloudVmRayBackend.
```

concretevitamin reviewed (commented):
Thanks; did a pass over the new changes.

Minor: Can we make the following errors for various CLI calls return exit code 1?
```
» AWS_PROFILE=AdministratorAccess-1234 sky start jump3                                       
Restarting 1 cluster: jump3. Proceed? [Y/n]:
Cluster 'jump3' (AWS) is owned by account 'yyyyy', but the currently activated account is 'xxxxx'.
```

concretevitamin commented:
> The problem may not be related to this PR as we don't change the generated ray yaml for the cluster. Do you think it is because the `jump` branch contains some additional settings in the ray yaml that is different from the one generated by the current branch causing the hash difference and resource leakage?

That's right! Thanks.

---

Potential issue

(1) Intentionally passing a wrong profile string `AdministratorAccess-1234`
```
» AWS_PROFILE=AdministratorAccess-1234 sky queue
Fetching and parsing job queue...
getting the job queue cluster 'jump3' (status: STOPPED)... skipped.
getting the job queue cluster 'smoke' (status: STOPPED)... skipped.

W 12-18 17:53:52 backend_utils.py:1900] Failed to refresh the cluster status, it is not fatal, but getting the job queue cluster 'sky-spot-controller-xxx' might hang if the cluster is not up.
W 12-18 17:53:52 backend_utils.py:1900] Detailed reason: Failed to get AWS user identity with unknown exception: <class 'botocore.exceptions.ProfileNotFound'> The config profile (AdministratorAccess-1234) could not be found.
getting the job queue cluster 'sky-spot-controller-xxx' (status: STOPPED)... skipped.
```

1. It's surprising this warning is printed for spot controller but not for the other two clusters. Is it needed?
2. (Related to an existing comment) "but getting the job queue cluster" - grammar is tricky.

(2) Passing the correct profile string `<correct>`
```
» AWS_PROFILE=<correct> sky queue                                                  1 ↵
Fetching and parsing job queue...
getting the job queue cluster 'jump3' (status: STOPPED)... skipped.
getting the job queue cluster 'smoke' (status: STOPPED)... skipped.

W 12-18 17:55:09 backend_utils.py:1900] Failed to refresh the cluster status, it is not fatal, but getting the job queue cluster 'sky-spot-controller-xxx' might hang if the cluster is not up.
W 12-18 17:55:09 backend_utils.py:1900] Detailed reason: Cluster 'sky-spot-controller-xxx' (AWS) is owned by account 'yyyy', but the currently activated account is 'zzzz'.
getting the job queue cluster 'sky-spot-controller-xxx' (status: STOPPED)... skipped.
```

---

Aside: I now think that carefully allowing certain operations on a non-owned cluster may not be worth the effort. This is because it's harder for us to get right or for users to remember the logic. On the other hand,  if it's clear cut the user doesn't need to learn new concepts: if one doesn't own a cluster, no operations are allowed. That said, let's ship it and see what user feedback is ;)

Michaelvll commented on sky/spot/spot_utils.py:
```diff
@@ -507,3 +507,37 @@ def load_job_table_cache() -> Tuple[str, str]:
+        logger.warning(
+            f'Failed to get the status of the spot controller. '
+            'It is not fatal, but spot commands may hang, when '
+            'the controller is not up, or have inaccurate statement.\n'
```
For 1, the inaccurate statement means the caller can print out "all the spot jobs are finished" or claim the spot controller in some status when it is not.

For 2, it can hang for about 10 seconds when trying to ssh into a stopped controller.

Michaelvll commented on sky/spot/spot_utils.py:
```diff
@@ -507,3 +507,37 @@ def load_job_table_cache() -> Tuple[str, str]:
+def is_spot_controller_up(
+    stopped_message: str,
+) -> Tuple[Optional[global_user_state.ClusterStatus],
+           Optional[backends.Backend.ResourceHandle]]:
```
Done. Thanks!

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1839,14 +1855,87 @@ def refresh_cluster_status_handle(
+    cluster_name: str,
+    operation: str,
+    *,
+    expected_backend: Optional[Type[backends.Backend]] = None
```
Done for 1.

For 2, since the `execution.exec` uses this function without a expected backend, we need to have way to turn off the check for hte backend. I changed it to a bool to make it more specific and most of the callers will not need to pass the argument. Wdyt?

Michaelvll commented on sky/cli.py:
```diff
@@ -3052,7 +3052,7 @@ def spot_cancel(name: Optional[str], job_ids: Tuple[int], all: bool, yes: bool):
 
     """
 
-    _, handle = _is_spot_controller_up(
```
Done. Thanks!

Michaelvll commented on sky/core.py:
```diff
@@ -356,10 +355,23 @@ def autostop(
+        with ux_utils.print_exception_no_traceback():
+            e.message += (f'\n  auto{option_str} is only supported by backend: '
+                          f'{backends.CloudVmRayBackend.NAME}')
+            raise e from None
```
That is a good point! In that case, we may want to change the behavior of `ux_utils.print_exception_no_traceback` based on the entrypoint (CLI or programmatic API)

Michaelvll commented:
Thanks for the review @concretevitamin!

> Minor: Can we make the following errors for various CLI calls return exit code 1?

This is a good call, but a bit difficult to decide the boundary. Since our `sky start` can operate on multiple clusters, do you think we should set the exit code to 1 if one of the clusters fails due to the identity problem?

> It's surprising this warning is printed for spot controller but not for the other two clusters. Is it needed?

Is it because only the spot controller has the autostop set up? We will only check the identity when we need to query the cloud CLI. For those clusters without autostop set up, there is no need to check or warn since the operator should work. However, for the cluster with autostop set up, there is no guarantee that the following codes will work, i.e. the warning might be useful as a disclaimer. Wdyt?

> (Related to an existing comment) "but getting the job queue cluster" - grammar is tricky.

Modified the logs based on the suggestions in the previous comments. The logging for a non-UP cluster will look like the following:
```
sky logs min
sky.exceptions.ClusterNotUpError: Tailing logs: skipped for cluster 'min' (status: STOPPED). It is only allowed for UP clusters.
```

I agree that the user may find the boundary a bit complicated, and we can modify it based on the feedback. One thing that may be noted is that we are not making multi-account officially supported after this PR, as that requires more detailed edge case handling. Instead, I would rather regard this PR as a safeguard for the user trying to switch from one account to another entirely and stay with the new account. ; )

Michaelvll commented:
As discussed offline, we decide to ban all operations for the identity mismatch for now. The new behavior will be:
```
AWS_PROFILE=admin-user sky logs min
sky.exceptions.ClusterOwnerIdentityMismatchError: Cluster 'min' (AWS) is owned by account '679991763071', but the currently activated account is 'AROATWUGBKI524LKNFJEB:admin-user'.
```
<img width="1190" alt="image" src="https://user-images.githubusercontent.com/6753189/208354841-d058a9f2-7a23-4816-b8c7-a341f7bd7ba3.png">

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1876,46 +1876,55 @@ def refresh_cluster_status_handle(
         # can still be done by only using ssh, but the ssh can hang if the
         # cluster is not up (e.g., autostopped).
+
+        # We do not catch the exception for cloud identity checking for now, so that
+        # to disable all the operators on the cluster created by another user identity.
```
```suggestion
        # We do not catch the exception for cloud identity checking for now, in order
        # to disable all operations on clusters created by another user identity.
```

concretevitamin commented on sky/utils/common_utils.py:
```diff
@@ -291,3 +291,15 @@ def decode_payload(payload_str: str) -> Any:
+
+
+def class_fullname(cls):
+    """Get the full name of a class.
```
Nice! The class printing has been ongoing for a while.

nit: Maybe useful to include an example input/output value.

Michaelvll replied:
Added. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -513,24 +513,44 @@ def is_spot_controller_up(
-            'the controller is not up, or have inaccurate statement.\n'
-            f'Details: {e}')
+            'It is not fatal, but spot commands/calls may hang or make '
+            'inaccurate statement, when the controller is not up.\n'
```
RE: https://github.com/skypilot-org/skypilot/pull/1513#discussion_r1051685347

I feel this can be a bit hard to understand. How about changing to "or return stale information"?

Michaelvll replied:
Good point! Changed. Thanks!

concretevitamin commented on sky/core.py:
```diff
@@ -417,6 +405,10 @@ def queue(cluster_name: str,
         RuntimeError: if failed to get the job queue.
         sky.exceptions.ClusterNotUpError: the cluster is not up.
         sky.exceptions.NotSupportedError: the feature is not supported.
+        sky.exceptions.ClusterOwnerIdentityMismatchError: if the current user is
```
Nit: can we copy paste the Raises section to other callers of check_cluster_available()?

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/cli.py:
```diff
@@ -1476,12 +1476,15 @@ def queue(clusters: Tuple[str], skip_finished: bool, all_users: bool):
+            click.echo(
+                f'{colorama.Fore.YELLOW}Failed to get the job queue for '
+                f'cluster {cluster!r}.{colorama.Style.RESET_ALL}\n'
+                f'  Details: [{common_utils.class_fullname(e.__class__)}] {e}')
```
```suggestion
                f'  {e}')
```
Not feeling too strongly. Just that current output seems a bit verbose:
<img width="1585" alt="Screen Shot 2022-12-19 at 18 41 03" src="https://user-images.githubusercontent.com/592670/208407725-18b8775f-fb91-42a6-8dd0-d3ed4e516de0.png">

(Maybe it's due to the details are showing up in yellow as well?)

Michaelvll replied:
I removed the color and `Details:` at the beginning, and it now looks like the following:
<img width="1002" alt="image" src="https://user-images.githubusercontent.com/6753189/208496932-5025a66f-16ad-4334-9ad4-8ba3cc6e1f60.png">

concretevitamin replied:
LGTM

concretevitamin reviewed (approved):
LGTM, thanks for the great work @Michaelvll!

Michaelvll commented:
Tested (7874618):
- [x] `tests/run_smoke_tests.sh`
